File: 46. Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector <int> visited (nums.size(), false);
        vector<vector<int>> res;
        DFS(res, {}, nums, visited);
        return res;
    }
    void DFS(vector<vector<int>>& res, vector<int> path, vector<int>& nums, vector<int> visited)
    {
        if (path.size() == nums.size())
        {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < nums.size(); i++)
        {
            if (!visited[i])
            {
                visited[i] = true;
                path.push_back(nums[i]);
                DFS(res, path, nums, visited);
                path.pop_back();
                visited[i] = false;
            }
        }
    }
};

File: 46. Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        self.res = []
        visited = [False for i in xrange(len(nums))]
        self.DFS([], nums, visited)
        return self.res
        
    def DFS(self, path, nums, visited):
        if len(path) == len(nums):
            self.res += [path[:]]
            return
        for i in xrange(len(nums)):
            if not visited[i]:
                visited[i] = True
                self.DFS(path + [nums[i]], nums, visited)
                visited[i] = False
